src/blueair_api/device_aws.py

```diff
@@ -27,71 +28,82 @@ async def create_device(cls, api, uuid, name, mac, type_name, refresh=False):
 
     async def refresh(self):
         _LOGGER.debug(f"refreshing blueair device aws: {self}")
         info = await self.api.device_info(self.name_api, self.uuid)
-        sensor_data = convert_api_array_to_dict(info["sensordata"])
-        self.pm1 = safely_get_json_value(sensor_data, "pm1", int)
-        self.pm2_5 = safely_get_json_value(sensor_data, "pm2_5", int)
-        self.pm10 = safely_get_json_value(sensor_data, "pm10", int)
-        self.tVOC = safely_get_json_value(sensor_data, "tVOC", int)
-        self.temperature = safely_get_json_value(sensor_data, "t", int)
-        self.humidity = safely_get_json_value(sensor_data, "h", int)
-
-        self.name = safely_get_json_value(info, "configuration.di.name")
-        self.firmware = safely_get_json_value(info, "configuration.di.cfv")
-        self.mcu_firmware = safely_get_json_value(info, "configuration.di.mfv")
-        self.serial_number = safely_get_json_value(info, "configuration.di.ds")
-        self.sku = safely_get_json_value(info, "configuration.di.sku")
-
-        states = convert_api_array_to_dict(info["states"])
-        self.running = safely_get_json_value(states, "standby") is False
-        self.night_mode = safely_get_json_value(states, "nightmode", bool)
-        self.germ_shield = safely_get_json_value(states, "germshield", bool)
-        self.brightness = safely_get_json_value(states, "brightness", int)
-        self.child_lock = safely_get_json_value(states, "childlock", bool)
-        self.fan_speed = safely_get_json_value(states, "fanspeed", int)
-        self.fan_auto_mode = safely_get_json_value(states, "automode", bool)
-        self.filter_usage = safely_get_json_value(states, "filterusage", int)
-        self.wifi_working = safely_get_json_value(states, "online", bool)
-        self.wick_usage = safely_get_json_value(states, "wickusage", int)
-        self.wick_dry_mode = safely_get_json_value(states, "wickdrys", bool)
-        self.auto_regulated_humidity = safely_get_json_value(states, "autorh", int)
-        self.water_shortage = safely_get_json_value(states, "wshortage", bool)
+        # ir.parse_json(ir.Attribute, ir.query_json(info, "configuration.da"))
+        ds = ir.parse_json(ir.Sensor, ir.query_json(info, "configuration.ds"))
+        dc = ir.parse_json(ir.Control, ir.query_json(info, "configuration.dc"))
+
+        sensor_data = ir.SensorPack(info["sensordata"]).to_latest_value()
+
+        def getter(data_dict, decl_dict, key):
+            return data_dict.get(key) if key in decl_dict else NotImplemented
+
+        self.pm1 = getter(sensor_data, ds, "pm1")
+        self.pm2_5 = getter(sensor_data, ds, "pm2_5")
+        self.pm10 = getter(sensor_data, ds, "pm10")
+        self.tVOC = getter(sensor_data, ds, "tVOC")
+        self.temperature = getter(sensor_data, ds, "t")
+        self.humidity = getter(sensor_data, ds, "h")
+
+        self.name = ir.query_json(info, "configuration.di.name")
```

Comment:
AttributeType serves for two purposes: the obvious one is marking an attribute as NotImplemented-able. There is an additional purpose: everything coming out of refresh() is marked AttributeType, and we have a unittest facility that ensures all of them are tested in the test cases (the assert_all_checked context manager). If we break this convention then we lose that enforcement as well. If we value that safety in the tests, then using the AttributeType for these di attributes is the least amount of typing (or we add a new type to mark AlwaysImpementedAttributeType, e.g.).

Conceptually if a di attributes is missing, should we say it is Unavailable or NotImplemented (conceptually because I don't think it actually occurs)? I think maybe we should have marked such attribute NotImplemented because the device doesn't support that attribute in that case.

Reply:
I updated to return NotImplemented if a di field is missing. The unit test is updated to reflect this if you look at 2c8f922ef613b23f9a5556e816f48ab63934a44b